OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+/** \file
+* This file defines the  DataTable class, which is used by OpenSim to provide  *
+* an in-memory container for data access and manipulation.                     *
+*/
+
+#ifndef OPENSIM_DATA_TABLE_H_
+#define OPENSIM_DATA_TABLE_H_
+
+// Standard headers.
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+#include <string>
+#include <type_traits>
+#include <limits>
+// Non-standard headers.
+#include <SimTKcommon.h>
+
+
+namespace OpenSim {
+  class AbstractDataTable;
+  template<typename ET> class DataTable_;
+
+  using DataTable = DataTable_<SimTK::Real>;
+
+  /// Enum to specify if the the input iterator is traversing data row-wise or
+  /// col-wise.
+  /// clang3.6 crashes if this is turned to a "enum class". Until it is fixed,
+  /// use a pre c++11 enum.
+  enum InputItDir{
+    ROWWISE, 
+    COLWISE
+  };
+
+  /// Bind two DataTables by row.
+  template<typename ET>
+  DataTable_<ET> rbindDataTables(const DataTable_<ET>& dt1, 
+                                 const DataTable_<ET>& dt2);
+
+  /// Bind two DataTables by col.
+  template<typename ET>
+  DataTable_<ET> cbindDataTables(const DataTable_<ET>& dt1, 
+                                 const DataTable_<ET>& dt2);
+
+  // Exceptions.
+  class ColumnDoesNotExist;
+  class ColumnHasLabel;
+  class ColumnHasNoLabel;
+  class ZeroElements;
+  class InvalidEntry;
+  class MetaDataKeyExists;
+  class MetaDataKeyDoesNotExist;
+  class MetaDataTypeMismatch;
+} // namespace OpenSim
+
+
+/**----------------------------------------------------------------------------*
+* AbstractDataTable is the base-class of all DataTable_ allowing storage of    *
+* DataTable_ of different types to be stored in containers. See Datatable_ for *
+* details on the interface.                                                    *
+*-----------------------------------------------------------------------------*/
+class OpenSim::AbstractDataTable {
+public:
+  AbstractDataTable() = default;
+  AbstractDataTable(const AbstractDataTable&) = delete;
+  AbstractDataTable(AbstractDataTable&&) = delete;
```

Comment:
I am closing this PR to start a new PR for next round of review. Want to make a final comment.

Firstly, I have changed the DataTable implementation to use ClonePtr but ClonePtr itself will need a move constructor for this to be efficient. DataTable's copy/move constructors and copy/move assignment operators are now defaulted.

Secondly, just to confirm, default copy/move constructor for base class can be deleted if the derived class explicitly defines copy/move constructor and the explicitly defined copy/move constructor does not want to call the base class's copy/move constructor explicitly. In the absence of explicit initialization of base class object in the member initialization list of the derived class's constructors, the compiler will automatically insert code to "default construct" the base class sub-object. Thus there is no need for base class's copy/move constructor.
If the derived class chooses to use compiler generated copy/move constructors, which implicitly call corresponding copy/move constructor in base class to initialize the base class sub-object, then base class needs not-deleted copy/move constructors.
What this means is that in my previous implementation (without ClonePtr), clone() (referred to as virtual constructor in my previous comment) function will work (and did work) fine. In my current implementation (with ClonePtr & default copy/move constructors for DataTable), AbstractDataTable needs not-deleted copy/move constructors.
For more details, refer to C++ standard -- clause 12.6.2 paragraph 9.